File: src/main/java/com/leet/hard/NQueens.rs
struct Solution {}

impl Solution {
  pub fn solve_n_queens(n: i32) -> Vec<Vec<String>> {
    fn render(board: &Vec<usize>, size: usize) -> Vec<String> {
      board
        .iter()
        .map(|&pos| ".".repeat(pos) + "Q" + &".".repeat(size - pos - 1))
        .collect()
    }

    fn is_valid(board: &Vec<usize>, row: usize, col: usize) -> bool {
      for i in 0..row {
        let diff: usize;
        if col < board[i] {
          diff = board[i] - col;
        } else {
          diff = col - board[i];
        }
        if board[i] == col || (row - i) == diff {
          return false;
        }
      }
      true
    }

    fn back_track(size: usize, row: usize, board: &mut Vec<usize>, res: &mut Vec<Vec<String>>) {
      if row >= size {
        res.push(render(board, size));
      } else {
        for col in 0..size {
          if is_valid(board, row, col) {
            board[row] = col;
            back_track(size, row + 1, board, res);
          }
        }
      }
    }

    let mut board = vec![0 as usize; n as usize];
    println!("{:?}", board);
    let mut res = vec![];
    back_track(n as usize, 0, &mut board, &mut res);
    res
  }
}

fn main() {
  let res = Solution::solve_n_queens(4);
  println!("{:?}", res);
}

File: src/main/java/com/leet/hard/NQueens.ts
function render(board: number[]): string[] {
	return board.map(row => {
		return ".".repeat(row) + "Q" + ".".repeat(board.length - row - 1);
	});
}

function isValid(board: number[], row: number, col: number): boolean {
	for (let i = 0; i < row; i++) {
		if (board[i] === col || (row - i) === Math.abs(col - board[i])) {
			return false;
		}
	}
	return true;
}

function solve(size: number, row: number, board: number[], ans: string[][]) {
	if (row >= size) {
		ans.push(render(board));
	} else {
		for (let col = 0; col < size; col++) {
			if (isValid(board, row, col)) {
				board[row] = col;
				solve(size, row + 1, board, ans);
			}
		}
	}
}

function solveNQueens(n: number): string[][] {
	const ans: string[][] = [];
	const board: number[] = [];
	solve(n, 0, board, ans);
	return ans;
};